### PR TITLE
TagBox - minimize dataSource loads count on item selection (T888848)

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

File: js/ui/tag_box.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import $ from '../core/renderer';
 import devices from '../core/devices';
 import dataUtils from '../core/element_data';
+const typeUtils = require('../core/utils/type');
```
Comment (ksercs) [inline js/ui/tag_box.js]: Shouldn't we change `require` to `import`?

File: js/ui/tag_box.js [hunk 2]
```diff
@@ -864,24 +865,66 @@ const TagBox = SelectBox.inherit({
 
     _renderTags: function() {
         const d = new Deferred();
+        let plainDataUsed = false;
```
Comment (ksercs) [inline js/ui/tag_box.js]: Don't u think, it's good idea to use `is` prefix for boolean vars? like `isPlainDataUsed`

File: js/ui/tag_box.js [hunk 3]
```diff
@@ -864,24 +865,66 @@ const TagBox = SelectBox.inherit({
...
+    _shouldGetDataFromPlain: function(values) {
+        return values && this._dataSource.isLoaded() && !this._wasSearch() && values.length <= this._getPlainItems().length;
+    },
+
+    _getItemsFromPlain: function(values) {
+        const plainItems = this._getPlainItems();
+        const selectedItems = plainItems.filter((dataItem) => {
+
+            for(let i = 0; i < values.length; i++) {
+                if(typeUtils.isObject(values[i])) {
```
Comment (ksercs) [inline js/ui/tag_box.js]: Let's create new variable for `values[i]`?

File: js/ui/tag_box.js [hunk 4]
```diff
@@ -864,24 +865,66 @@ const TagBox = SelectBox.inherit({
...
+                    if(this._isValueEquals(dataItem, values[i])) {
+                        return true;
+                    }
+                } else {
```
Comment (ksercs) [inline js/ui/tag_box.js]: Let's use `else if` here?

File: js/ui/tag_box.js [hunk 5]
```diff
@@ -864,24 +865,66 @@ const TagBox = SelectBox.inherit({
 
     _renderTags: function() {
         const d = new Deferred();
+        let plainDataUsed = false;
 
-        this._loadTagsData().always((items) => {
-            if(this._disposed) {
-                d.reject();
-                return;
+        if(this._shouldGetDataFromPlain(this._valuesToUpdate)) {
```
Comment (ksercs) [inline js/ui/tag_box.js]: Let's make the names `_shouldGetDataFromPlain` and `_getItemsFromPlain` more consistent? etc   `_shouldGetItemsFromPlain`

File: js/ui/text_box/ui.text_editor.base.js [hunk 6]
```diff
@@ -710,7 +710,7 @@ const TextEditorBase = Editor.inherit({
             case 'text':
                 break;
             case 'value':
-                this._updateValue();
+                this._updateValue(this._useValuesToUpdate);
```
Comment (DokaRus) [inline js/ui/text_box/ui.text_editor.base.js]: Why do we need this?

```suggestion
                this._updateValue();
```

File: js/ui/tag_box.js [hunk 7]
```diff
@@ -549,8 +550,8 @@ const TagBox = SelectBox.inherit({
         this._renderPreventBlur(this._inputWrapper());
     },
 
-    _renderInputValueImpl: function() {
-        return this._renderMultiSelect();
+    _renderInputValueImpl: function(args) {
+        return this._renderMultiSelect(args);
```
Comment (DokaRus) [inline js/ui/tag_box.js]: Do we need this `args`?

```js
_renderInputValueImpl: function() {
        return this._renderMultiSelect();
```

File: js/ui/tag_box.js [hunk 8]
```diff
@@ -1389,7 +1432,9 @@ const TagBox = SelectBox.inherit({
                 this._setListOption('selectAllText', this.option('selectAllText'));
                 break;
             case 'value':
+                this._valuesToUpdate = args?.value;
```
Comment (DokaRus) [inline js/ui/tag_box.js:1435]:
```suggestion
                this._valuesToUpdate = args.value;
```

**Review by DokaRus — APPROVED** [review]

**Review by ksercs — APPROVED** [review]

**Review by San4es — COMMENTED** [review]
LGTM

**Pull request merged**